illumos-utils/src/opte/non_illumos.rs

```diff
@@ -43,3 +70,268 @@ pub fn delete_all_xde_devices(log: &Logger) -> Result<(), Error> {
+            != IpCidr::Ip4(Ipv4Cidr::new(
+                Ipv4Addr::ANY_ADDR,
+                0.try_into().unwrap(),
+            ))
+        {
+            return false;
+        }
+        if !matches!(self.target, RouterTarget::InternetGateway(None)) {
+            return false;
+        }
+        matches!(self.class, RouterClass::System)
+    }
+}
+
+impl From<&AddRouterEntryReq> for RouteInfo {
+    fn from(value: &AddRouterEntryReq) -> Self {
+        Self { dest: value.dest, target: value.target, class: value.class }
+    }
+}
+impl From<&DelRouterEntryReq> for RouteInfo {
+    fn from(value: &DelRouterEntryReq) -> Self {
+        Self { dest: value.dest, target: value.target, class: value.class }
+    }
+}
+
+/// Data for one OPTE port
+#[derive(Debug)]
+pub(crate) struct PortData {
+    /// The OPTE-layer information
+    #[cfg_attr(test, allow(dead_code))]
+    pub port: PortInfo,
+    /// The routes for this port. This simulates the router layer.
+    pub routes: Vec<RouteInfo>,
+}
+
+#[derive(Debug)]
+pub(crate) struct State {
+    pub ports: HashMap<String, PortData>,
+    pub underlay_initialized: bool,
+}
+
+const NO_RESPONSE: NoResp = NoResp { unused: 99 };
+static OPTE_STATE: OnceLock<Mutex<State>> = OnceLock::new();
+
+fn opte_state() -> &'static Mutex<State> {
+    OPTE_STATE.get_or_init(|| {
+        Mutex::new(State { ports: HashMap::new(), underlay_initialized: true })
+    })
+}
+
+/// Simulated handle to OPTE
+pub struct Handle;
+
+impl Handle {
+    /// Create a new handle to the test OPTE kernel module.
+    pub fn new() -> Result<Self, OpteError> {
+        Ok(Self)
+    }
+
+    /// Helper to get the OPTE state.
+    #[cfg(test)]
+    pub(crate) fn state(&self) -> &'static Mutex<State> {
+        opte_state()
+    }
+
+    /// Add a new port.
+    pub fn create_xde(
+        &self,
+        name: &str,
+        cfg: VpcCfg,
+        _: DhcpCfg,
+        _: bool,
+    ) -> Result<NoResp, OpteError> {
+        let name = name.to_string();
+        let IpCfg::Ipv4(ip_cfg) = cfg.ip_cfg else {
+            unimplemented!("IPv6 support");
+        };
+        let ephemeral_ip4_addr =
+            ip_cfg.external_ips.snat.as_ref().map(|snat| snat.external_ip);
+        let port = PortInfo {
+            name: name.clone(),
+            mac_addr: cfg.guest_mac,
+            ip4_addr: Some(ip_cfg.private_ip),
+            ephemeral_ip4_addr,
+            floating_ip4_addrs: None,
+            ip6_addr: None,
+            ephemeral_ip6_addr: None,
+            floating_ip6_addrs: None,
+            state: "created".to_string(),
+        };
+        let mut state = opte_state().lock().unwrap();
+        anyhow::ensure!(
+            state.underlay_initialized,
+            "Underlay is not initialized"
+        );
+        assert!(
+            state
+                .ports
+                .insert(name, PortData { port, routes: Vec::new() })
+                .is_none(),
+            "duplicate OPTE ports"
+        );
```

Comment:
This seems fine but out of curiosity - why `assert!` here when similar errors result in `bail!`s (e.g., adding a router entry to a nonexistent port)?

Reply:
Bailing out seems good too, will switch.

Reply:
Done in 29b8e7b4d